core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -0,0 +1,671 @@
+    return getObjectStatus(stripPrefixIfPresent(path)) != null;
+  }
+
+  @Override
+  public UnderFileStatus[] listStatus(String path) throws IOException {
+    return listInternal(path, ListOptions.defaults());
+  }
+
+  @Override
+  public UnderFileStatus[] listStatus(String path, ListOptions options)
+      throws IOException {
+    return listInternal(path, options);
+  }
+
+  @Override
+  public boolean mkdirs(String path, MkdirsOptions options) throws IOException {
+    if (path == null) {
+      return false;
+    }
+    if (isDirectory(path)) {
+      return true;
+    }
+    if (isFile(path)) {
+      LOG.error("Cannot create directory {} because it is already a file.", path);
+      return false;
+    }
+    if (!options.getCreateParent()) {
+      if (parentExists(path)) {
+        // Parent directory exists
+        return mkdirsInternal(path);
+      } else {
+        LOG.error("Cannot create directory {} because parent does not exist", path);
+        return false;
+      }
+    }
+    // Parent directories should be created
+    if (parentExists(path)) {
+      // Parent directory exists
+      return mkdirsInternal(path);
+    } else {
+      String parentKey = getParentPath(path);
+      // Recursively make the parent folders
+      return mkdirs(parentKey) && mkdirsInternal(path);
+    }
+  }
+
+  @Override
+  public InputStream open(String path, OpenOptions options) throws IOException {
+    return openObject(stripPrefixIfPresent(path), options);
+  }
+
+  @Override
+  public boolean renameDirectory(String src, String dst) throws IOException {
+    UnderFileStatus[] children = listInternal(src, ListOptions.defaults());
+    if (children == null) {
+      LOG.error("Failed to list directory {}, aborting rename.", src);
+      return false;
+    }
+    if (isFile(dst) || isDirectory(dst)) {
+      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
+      return false;
+    }
+    // Source exists and is a file, and destination does not exist
+    // Rename the source folder first
+    if (!copyObject(stripPrefixIfPresent(convertToFolderName(src)),
+        stripPrefixIfPresent(convertToFolderName(dst)))) {
+      return false;
+    }
+    // Rename each child in the src folder to destination/child
+    for (UnderFileStatus child : children) {
+      String childSrcPath = PathUtils.concatPath(src, child);
+      String childDstPath = PathUtils.concatPath(dst, child);
+      boolean success;
+      if (child.isDirectory()) {
+        // Recursive call
+        success = renameDirectory(childSrcPath, childDstPath);
+      } else {
+        success = renameFile(childSrcPath, childDstPath);
+      }
+      if (!success) {
+        LOG.error("Failed to rename path {}, aborting rename.", child);
+        return false;
+      }
+    }
+    // Delete src and everything under src
+    return deleteDirectory(src, DeleteOptions.defaults().setRecursive(true));
```

Comment:
Not sure if there is any overhead, but I think we can get away with only deleting the direct children.